### PR TITLE
Adds support for custom types

### PR DESCRIPTION
**Pull request by felipecsl**
@gabrielittner this is my first draft on custom types. If you're happy with it, it's pretty easy to later add support to Enums by implementing an `EnumFactory` for it.
I hate the name `@AutoValueCursorFactory` but can't think of a better one. I'm open for ideas.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,6 +33,62 @@ import com.gabrielittner.auto.value.cursor.ColumnName;
...
+    return AutoValue_User.createFromCursor(cursor);
+  }
+}
+```
+
+AvatarFactory.java
+
+```java
+public class AvatarFactory {
+  public static Avatar(Cursor cursor) {
```
Comment (gabrielittner) [inline README.md]: the method name is missing

**Comment by gabrielittner** [comment]
It's probably the best if the annotation name for this and the `auto-value-parcel` adapter follow the same convention. It is simply using `@ParcelAdapter`, but `@CursorAdapter` clashes with `android.widget.CursorAdapter`.

cc @rharter

**Comment by rharter** [comment]
@gabrielittner I think that makes  a lot of sense.  Also, in this case, the factory isn't creating an `AutoValueCursor`, but just a `Cursor`.  

I understand the concern about the conflict with `android.widget.CursorAdapter`, but I would point out that it seems extremely unlikely that they'd ever be used in the same place, so as to actually conflict.  Still makes an annoyance for auto complete.  There's also the fact that one is an annotation, and another is not, so autocomplete might be smart enough to differentiate the two.

Another option could be to change both to `TypeAdapter`s.  So we'd have `@ParcelTypeAdapter` and `@CursorTypeAdapter`.  That might make things a little more clear.

**Comment by gabrielittner** [comment]
In regards to the conflict, I meant that it could be confusing to someone. Using `TypeAdapter` sounds good to me.

File: README.md [hunk 2]
```diff
@@ -33,6 +33,62 @@ import com.gabrielittner.auto.value.cursor.ColumnName;
...
+ * `short`/`Short`
+ * `String`
+ * `boolean`/`Boolean`
+
+For other types, you need to use the `@AutoValueCursorFactory` annotation and specify a factory
+class that will be used to construct the type from the `Cursor` object. By convention, this factory
+class needs to have a public static method named `create()` that takes the `Cursor` and returns the
+custom type. Eg.:
+
+User.java
```
Comment (JakeWharton) [inline README.md]:
```
`User.java`:
```

File: README.md [hunk 3]
```diff
@@ -33,6 +33,62 @@ import com.gabrielittner.auto.value.cursor.ColumnName;
...
+  abstract String name();
+  @AutoValueCursorFactory(AvatarFactory.class) Avatar avatar();
+
+  public static User create(Cursor cursor) {
+    return AutoValue_User.createFromCursor(cursor);
+  }
+}
+```
+
+AvatarFactory.java
```
Comment (JakeWharton) [inline README.md]:
```
`AvatarFactory.java`:
```

File: README.md [hunk 4]
```diff
@@ -33,6 +33,62 @@ import com.gabrielittner.auto.value.cursor.ColumnName;
...
+public class AvatarFactory {
+  public static Avatar(Cursor cursor) {
+    String smallImageUrl = cursor.getString(cursor.getColumnIndex("small_image_url");
+    String largeImageUrl = cursor.getString(cursor.getColumnIndex("large_image_url");
+    return new Avatar(smallImageUrl, largeImageUrl);
+  }
+}
+```
+
+Avatar.java
```
Comment (JakeWharton) [inline README.md]:
```
`Avatar.java`:
```

File: auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java [hunk 5]
```diff
@@ -124,21 +111,25 @@ private MethodSpec createReadMethod(Context context, String className,
...
-
             String cursorMethod = getCursorMethod(type);
             String columnName = getColumnName(element);
+            TypeMirror factoryTypeMirror = getFactoryTypeMirror(element);
             if (cursorMethod != null) {
                 readMethod.addStatement(cursorMethod, type, name,
                         columnName != null ? columnName : name);
+            } else if (factoryTypeMirror != null) {
+                TypeName factoryType = TypeName.get(factoryTypeMirror);
+                readMethod.addStatement("$T $N = $T.create(cursor)", type, name, factoryType);
```
Comment (gabrielittner) [inline auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java]: It's nicer to stay flexible on the naming of the factory method. The factory could just be another auto value class and the extension only looks at return type and parameter in `applicable()`. 

As an example:
```java
@AutoValue public abstract class User {
  abstract String id();
  abstract String name();
  @AutoValueCursorFactory(Avatar.class) Avatar avatar();

  public static User create(Cursor cursor) {
    return AutoValue_User.createFromCursor(cursor);
  }
}

@AutoValue public abstract class Avatar {
  public abstract String smallImageUrl();
  public abstract String largeImageUrl();

  public Avatar fromCursor(Cursor cursor) {
    return AutoValue_Avatar.createFromCursor(cursor);
  }
}
```

Comment (felipecsl) [inline auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java]: ok that's actually a great idea. Can save you a class

**Comment by felipecsl** [comment]
Alright, PR updated based on the feedback, please take another look.
Also heads up that apparently the AutoValue snapshot has changed in the last 2 days and it no longer allows `@Nullable` annotation on primitive fields. I updated the `types()` test to fix the failure by not using primitive types.

File: auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java [hunk 6]
```diff
@@ -221,6 +223,53 @@ private static boolean hasAnnotationWithName(Element element, String simpleName)
...
+                    || !element.getModifiers().contains(Modifier.STATIC)) {
+                continue;
+            }
+            ExecutableElement method = (ExecutableElement) element;
+            // that method should return the annotated class and take a Cursor as parameter
+            if (methodTakesAndReturns(method, CURSOR, returnType)) {
+                return method.getSimpleName().toString();
+            }
+        }
+        return null;
```
Comment (gabrielittner) [inline auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java]: Throw here with an appropriate error message instead of returning null. Right now javapoet would throw an IllegalArgumentException in line 132 for passing in null.

**Comment by gabrielittner** [comment]
When that last thing is changed and the commits are squashed we're good to merge-

File: auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java (hunk 6 above)
Comment (felipecsl) [inline auto-value-cursor/src/main/java/com/gabrielittner/auto/value/cursor/AutoValueCursorExtension.java]: good point, I noticed this when testing as well, hard to parse the error

**Comment by felipecsl** [comment]
alright sir, fixed up and squashed :+1:

**Comment by gabrielittner** [comment]
Thanks for the contribution :+1:
